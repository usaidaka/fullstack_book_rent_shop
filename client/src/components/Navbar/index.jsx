import PropTypes from 'prop-types';
import { useState } from 'react';
import { connect, useDispatch } from 'react-redux';
import { setLocale, setTheme } from '@containers/App/actions';

import * as React from 'react';
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import CssBaseline from '@mui/material/CssBaseline';
import Divider from '@mui/material/Divider';
import Drawer from '@mui/material/Drawer';
import IconButton from '@mui/material/IconButton';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemIcon from '@mui/material/ListItemIcon';
import { Link, useNavigate } from 'react-router-dom';
import MenuIcon from '@mui/icons-material/Menu';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import { Avatar, Menu, MenuItem } from '@mui/material';
import { FormattedMessage } from 'react-intl';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import LightModeIcon from '@mui/icons-material/LightMode';
import NightsStayIcon from '@mui/icons-material/NightsStay';
import { logout } from '@utils/logout';
import { createStructuredSelector } from 'reselect';
import { selectLogin, selectUser } from '@containers/Client/selectors';
import config from '@config/index';

/* ICONS */
import ExitToAppIcon from '@mui/icons-material/ExitToApp';
import HowToRegIcon from '@mui/icons-material/HowToReg';

import logoBook from '../../assets/logoBook.png';
import classes from './style.module.scss';

const drawerWidth = 240;

const Navbar = ({ locale, theme, children, login, user }) => {
  console.log(login);
  console.log(user);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const [menuPosition, setMenuPosition] = useState(null);

  const open = Boolean(menuPosition);
  const [mobileOpen, setMobileOpen] = React.useState(false);
  const [isClosing, setIsClosing] = React.useState(false);

  const handleDrawerClose = () => {
    setIsClosing(true);
    setMobileOpen(false);
  };

  const handleClick = (event) => {
    setMenuPosition(event.currentTarget);
  };

  const handleDrawerTransitionEnd = () => {
    setIsClosing(false);
  };

  const handleDrawerToggle = () => {
    if (!isClosing) {
      setMobileOpen(!mobileOpen);
    }
  };

  const handleClose = () => {
    setMenuPosition(null);
  };

  const onSelectLang = (lang) => {
    if (lang !== locale) {
      dispatch(setLocale(lang));
    }
    handleClose();
  };

  const handleTheme = () => {
    dispatch(setTheme(theme === 'light' ? 'dark' : 'light'));
  };

  const handleLogout = () => {
    logout(dispatch, navigate);
  };

  const drawer = (
    <div>
      <Toolbar />
      <Divider />
      <Link to="/admin/register-admin">
        <List onClick={handleLogout}>
          <ListItem disablePadding>
            <ListItemButton>
              <ListItemIcon>
                <HowToRegIcon />
              </ListItemIcon>
              <FormattedMessage id="registerAdmin" />
            </ListItemButton>
          </ListItem>
        </List>
      </Link>
      <Divider />
      <List onClick={handleLogout}>
        <ListItem disablePadding>
          <ListItemButton>
            <ListItemIcon>
              <ExitToAppIcon />
            </ListItemIcon>
            <FormattedMessage id="logout" />
          </ListItemButton>
        </ListItem>
      </List>
    </div>
  );

  return (
    <Box sx={{ display: 'flex' }}>
      <CssBaseline />
      <AppBar
        position="fixed"
        sx={{
          width: { sm: `calc(100% - ${drawerWidth}px)` },
          ml: { sm: `${drawerWidth}px` },
        }}
      >
        <Toolbar>
          <IconButton
            color="inherit"
            aria-label="open drawer"
            edge="start"
            onClick={handleDrawerToggle}
            sx={{ mr: 2, display: { sm: 'none' } }}
          >
            <MenuIcon />
          </IconButton>
          <Box className={classes.headerWrapper}>
            <Box className={classes.header}>
              <img src={logoBook} alt="" width={40} />
              <Typography variant="h6" noWrap component="div">
                <FormattedMessage id="app_title_header" />
              </Typography>
            </Box>
            <div className={classes.contentWrapper}>
              <div className={classes.toolbar}>
                <div className={classes.theme} onClick={handleTheme} data-testid="toggleTheme">
                  {theme === 'light' ? <NightsStayIcon /> : <LightModeIcon />}
                </div>
                <div className={classes.toggle} onClick={handleClick}>
                  <Avatar className={classes.avatar} src={locale === 'id' ? '/id.png' : '/en.png'} />
                  <div className={classes.lang}>{locale}</div>
                  <ExpandMoreIcon />
                </div>
              </div>
              <Menu open={open} anchorEl={menuPosition} onClose={handleClose}>
                <MenuItem onClick={() => onSelectLang('id')} selected={locale === 'id'}>
                  <div className={classes.menu}>
                    <Avatar className={classes.menuAvatar} src="/id.png" />
                    <div className={classes.menuLang}>
                      <FormattedMessage id="app_lang_id" />
                    </div>
                  </div>
                </MenuItem>
                <MenuItem onClick={() => onSelectLang('en')} selected={locale === 'en'}>
                  <div className={classes.menu}>
                    <Avatar className={classes.menuAvatar} src="/en.png" />
                    <div className={classes.menuLang}>
                      <FormattedMessage id="app_lang_en" />
                    </div>
                  </div>
                </MenuItem>
              </Menu>
              {/* profile */}
              <div>
                <div className={classes.toolbar}>
                  <div className={classes.toggle}>
                    <img
                      src={`${config.api.image_customer}${user.image}`}
                      alt=""
                      className={classes['photo-profile']}
                    />
                  </div>
                </div>
              </div>
            </div>
          </Box>
        </Toolbar>
      </AppBar>
      <Box component="nav" sx={{ width: { sm: drawerWidth }, flexShrink: { sm: 0 } }} aria-label="mailbox folders">
        {/* The implementation can be swapped with js to avoid SEO duplication of links. */}
        <Drawer
          variant="temporary"
          open={mobileOpen}
          onTransitionEnd={handleDrawerTransitionEnd}
          onClose={handleDrawerClose}
          ModalProps={{
            keepMounted: true, // Better open performance on mobile.
          }}
          sx={{
            display: { xs: 'block', sm: 'none' },
            '& .MuiDrawer-paper': { boxSizing: 'border-box', width: drawerWidth },
          }}
        >
          {drawer}
        </Drawer>
        <Drawer
          variant="permanent"
          sx={{
            display: { xs: 'none', sm: 'block' },
            '& .MuiDrawer-paper': { boxSizing: 'border-box', width: drawerWidth },
          }}
          open
        >
          {drawer}
        </Drawer>
      </Box>
      <Box component="main" sx={{ flexGrow: 1, p: 3, width: { sm: `calc(100% - ${drawerWidth}px)` } }}>
        <Toolbar />
        {children}
      </Box>
    </Box>
  );
};

Navbar.propTypes = {
  locale: PropTypes.string.isRequired,
  theme: PropTypes.string,
  children: PropTypes.element.isRequired,
  login: PropTypes.bool,
  user: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  login: selectLogin,
  user: selectUser,
});

export default connect(mapStateToProps)(Navbar);
