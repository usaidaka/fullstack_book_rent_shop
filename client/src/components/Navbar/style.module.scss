@import '@styles/base';
a,
a:visited,
a:hover,
a:active {
  color: inherit;
  text-decoration: none;
}

.active {
  background-color: $color-primary-header;
  color: white;
}

.logout {
  color: red;
  border-radius: 20px;
}

.headerWrapper {
  @include directionX;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 20px;

  .header {
    display: flex;
    gap: 15px;
  }

  .contentWrapper {
    @include directionX;

    color: var(--color-text);
    padding: 1rem;
    align-items: center;
    justify-content: space-between;
    font-size: 1rem;

    .logoImage {
      @include directionX;
      cursor: pointer;
      align-items: center;

      .logo {
        width: 1.5rem;
        height: 1.5rem;
        margin-right: 0.5rem;
      }
      .title {
        font-weight: 700;
      }
    }

    .toolbar {
      @include directionX;
      .theme {
        @include directionX;
        align-items: center;
        margin-right: 0.5rem;
        cursor: pointer;
      }
      .toggle {
        @include directionX;
        align-items: center;
        cursor: pointer;

        .photo-profile {
          width: 1.5rem;
          height: 1.5rem;
          border-radius: 100%;
        }

        .avatar {
          width: 1.2rem;
          height: 1.2rem;
          box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
          margin-right: 0.5rem;
        }

        .lang {
          font-family: Poppins, Arial, Helvetica, sans-serif;
          font-weight: 700;
          font-size: 1rem;
          text-transform: uppercase;
        }

        .arrow {
          font-size: 1rem;
        }
      }
    }
  }
}

.menu {
  @include directionX;
  align-items: center;

  .menuAvatar {
    width: 1rem;
    height: 1rem;
    box-shadow: rgba(99, 99, 99, 0.2) 0px 2px 8px 0px;
    margin-right: 0.5rem;
  }

  .menuLang {
    font-family: Poppins, Arial, Helvetica, sans-serif;
    font-size: 0.8rem;
  }
}
