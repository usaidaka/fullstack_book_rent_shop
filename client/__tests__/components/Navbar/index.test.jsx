import { render } from '@testing-library/react';

import Navbar from '@components/Navbar';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

describe('Navbar Component', () => {
  test('Correct render', () => {
    const navbar = render(<Navbar title="Title" locale="en" theme="light" />);
    expect(navbar.getByTestId('navbar')).toBeInTheDocument();
  });

  test('Should match with snapshot', () => {
    const navbar = render(<Navbar title="Title" locale="en" theme="light" />);
    expect(navbar).toMatchSnapshot();
  });
});
