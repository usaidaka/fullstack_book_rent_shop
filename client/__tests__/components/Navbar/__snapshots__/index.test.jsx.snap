// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="headerWrapper"
        data-testid="navbar"
      >
        <div
          class="contentWrapper"
        >
          <div
            class="logoImage"
          >
            <img
              alt="logo"
              class="logo"
              src="/vite.svg"
            />
            <div
              class="title"
            >
              Title
            </div>
          </div>
          <div
            class="toolbar"
          >
            <div
              class="theme"
              data-testid="toggleTheme"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="NightsStayIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11.1 12.08c-2.33-4.51-.5-8.48.53-10.07C6.27 2.2 1.98 6.59 1.98 12c0 .14.02.28.02.42.62-.27 1.29-.42 2-.42 1.66 0 3.18.83 4.1 2.15 1.67.48 2.9 2.02 2.9 3.85 0 1.52-.87 2.83-2.12 3.51.98.32 2.03.5 3.11.5 3.5 0 6.58-1.8 8.37-4.52-2.36.23-6.98-.97-9.26-5.41z"
                />
                <path
                  d="M7 16h-.18C6.4 14.84 5.3 14 4 14c-1.66 0-3 1.34-3 3s1.34 3 3 3h3c1.1 0 2-.9 2-2s-.9-2-2-2z"
                />
              </svg>
            </div>
            <div
              class="toggle"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular avatar css-1wlk0hk-MuiAvatar-root"
              >
                <img
                  class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                  src="/en.png"
                />
              </div>
              <div
                class="lang"
              >
                en
              </div>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ExpandMoreIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="headerWrapper"
      data-testid="navbar"
    >
      <div
        class="contentWrapper"
      >
        <div
          class="logoImage"
        >
          <img
            alt="logo"
            class="logo"
            src="/vite.svg"
          />
          <div
            class="title"
          >
            Title
          </div>
        </div>
        <div
          class="toolbar"
        >
          <div
            class="theme"
            data-testid="toggleTheme"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="NightsStayIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11.1 12.08c-2.33-4.51-.5-8.48.53-10.07C6.27 2.2 1.98 6.59 1.98 12c0 .14.02.28.02.42.62-.27 1.29-.42 2-.42 1.66 0 3.18.83 4.1 2.15 1.67.48 2.9 2.02 2.9 3.85 0 1.52-.87 2.83-2.12 3.51.98.32 2.03.5 3.11.5 3.5 0 6.58-1.8 8.37-4.52-2.36.23-6.98-.97-9.26-5.41z"
              />
              <path
                d="M7 16h-.18C6.4 14.84 5.3 14 4 14c-1.66 0-3 1.34-3 3s1.34 3 3 3h3c1.1 0 2-.9 2-2s-.9-2-2-2z"
              />
            </svg>
          </div>
          <div
            class="toggle"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular avatar css-1wlk0hk-MuiAvatar-root"
            >
              <img
                class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                src="/en.png"
              />
            </div>
            <div
              class="lang"
            >
              en
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
